core/server/worker/src/main/java/alluxio/worker/block/UfsIOManager.java

```diff
@@ -0,0 +1,298 @@
+            MetricsSystem.escape(key.getUri()))
+            : MetricsSystem.meterWithTags(MetricKey.WORKER_BYTES_READ_UFS.getName(),
+            MetricKey.WORKER_BYTES_READ_UFS.isClusterAggregated(), MetricInfo.TAG_UFS,
+            MetricsSystem.escape(key.getUri()), MetricInfo.TAG_USER, key.getUser()));
+    return ufsBytesReadThroughput.getOneMinuteRate();
+  }
+
+  /**
+   * read from ufs.
+   * 
+   * @param blockId block id
+   * @param blockSize block size
+   * @param ufsPath ufs path
+   * @param offset offset to read
+   * @param length length to read
+   * @param positionShort is position short or not
+   * @param user user tag
+   * @return content read
+   * @throws ResourceExhaustedException any exception from ufs
+   */
+  public CompletableFuture<byte[]> read(long blockId, long blockSize, String ufsPath, long offset,
+      long length, boolean positionShort, String user) throws ResourceExhaustedException {
+    CompletableFuture<byte[]> future = new CompletableFuture<>();
+    if (mReadQueue.size() >= READ_CAPACITY) {
+      throw new ResourceExhaustedException(format("UFS read at capacity, try later or increase %s",
+          PropertyKey.UNDERFS_IO_READ_QUEUE_CAPACITY.getName()));
+    }
+    mReadQueue.add(
+        new ReadTask(blockId, blockSize, ufsPath, offset, length, positionShort, user, future));
+    return future;
+  }
+
+  private class ReadTask implements Runnable {
+    private final long mOffset;
+    private final long mLength;
+    private final boolean mIsPositionShort;
+    private final CompletableFuture<byte[]> mFuture;
+    private final long mBlockId;
+    private final String mUfsPath;
+    private final String mUser;
+    private final long mBlockSize;
+
+    private ReadTask(long blockId, long blockSize, String ufsPath, long offset, long length,
+        boolean positionShort, String user, CompletableFuture<byte[]> future) {
+      mUser = user;
+      mBlockId = blockId;
+      mBlockSize = blockSize;
+      mUfsPath = ufsPath;
+      mOffset = offset;
+      mLength = length;
+      mIsPositionShort = positionShort;
+      mFuture = future;
+    }
+
+    public String getUser() {
+      return mUser;
+    }
+
+    public void run() {
+      try {
+        byte[] buffer = readInternal();
+        mFuture.complete(buffer);
+      } catch (IOException e) {
+        mFuture.completeExceptionally(e);
+      }
+    }
+
+    public byte[] readInternal() throws IOException {
+      long bytesToRead = Math.min(mLength, mBlockSize - mOffset);
+      if (bytesToRead <= 0) {
+        return new byte[0];
+      }
+      UnderFileSystem ufs = mUfsClient.acquireUfsResource().get();
+      long blockStart = BlockId.getSequenceNumber(mBlockId) * mBlockSize;
+      byte[] data = new byte[(int) bytesToRead];
+      int bytesRead = 0;
+      InputStream inStream = null;
+      try {
+        inStream = mUfsInstreamCache.acquire(ufs, mUfsPath, IdUtils.fileIdFromBlockId(mBlockId),
```

Comment:
I read a little bit of the `UfsInputStreamCache`. I don't know why we should use it here. Reading the PR, it's trying to avoid repeatedly creating the same stream, but I don't think we'll do that anyways, so why paying the overhead? If you read the implementation, this cache is pretty expensive to maintain.

Reply:
My understanding is that for non Seekable instream the overhead is pretty low. And for the seekable stream, we can reuse the stream for the same file(for example load different blocks within the file without reopening the stream again and again).